feat/implementing change-of-day, tasks info and clickable blocks

aizensoosuke opened the pull request:
I also separated `org-timeline--generate-hourline` from within `org-timeline--generate-timeline`.

Commit:
feat: implement configurable "midnight/change-of-day"

note: the list of tasks and the timeline are now generated in separate functions.

Co-authored-by: Matus Goljer <dota.keys@gmail.com>

aizensoosuke commented on org-timeline.el:
```diff
@@ -158,22 +158,21 @@ Return new copy of STRING."
           (insert timeline)
           (-each tasks
-            (-lambda ((beg end face))
-              (while (get-text-property (get-start-pos current-line beg) 'org-timeline-occupied)
```
Just swapping those two lines

Fuco1 commented on org-timeline.el:
```diff
@@ -116,20 +121,11 @@ Return new copy of STRING."
-  "Generate the timeline string that will represent current agenda view."
-  (let* ((start-offset 300)
+(defun org-timeline--list-tasks ()
+  "Builds the list of tasks to display"
```
The docstrings should use imperative and end with a dot

```suggestion
  "Build the list of tasks to display."
```

Fuco1 commented on org-timeline.el:
```diff
@@ -146,27 +142,56 @@ Return new copy of STRING."
+              (push (list beg end face) tasks))))))
+    (nreverse tasks)))
+
+(defun org-timeline--generate-hourline ()
+  "Generate the template for the timeline's hourline."
+  (let* ((hourline "|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|")
+         (start-hour (abs org-timeline-start-hour))
+         (current-slot 0))
+    (while (< current-slot 24)
+      (let* ((current-slot-hour (mod (+ start-hour current-slot) 24))
+            (current-slot-ones (mod current-slot-hour 10))
+            (current-slot-tens (/ (- current-slot-hour current-slot-ones) 10)))
+        (aset hourline (+ 1 (* current-slot 6)) (+ 48 current-slot-tens)) ;; 48 is ascii for '0'
+        (aset hourline (+ 2 (* current-slot 6)) (+ 48 current-slot-ones)))
+      (cl-incf current-slot))
+    hourline))
```
Usually using `aset` and similar functions which modify by side effect is not the common approach in lisp.  More lispy/functional approach would be to generate each hour slot and then join them into one string.

A possible implementation could be something like this.  Of course there are some additional changes needed, like the wrapping `|` and customizable start (currently I just put 5 there).

```elisp
(mapconcat (lambda (x) (format "%02d:00" (mod x 24))) (number-sequence 5 (+ 5 23)) "|")
```

This works by first creating a number sequence, then applying the lambda on each item of the list and finally joining all the strings with `"|"`.

It could also be implemented in two steps, first using `mapcar` to generate the list:

```elisp
(mapcar (lambda (x) (format "%02d:00" (mod x 24))) (number-sequence 5 (+ 5 23)))
```

and then join it into one string

```elisp
(string-join list-of-hours "|")
```

aizensoosuke replied:
Thank you for the tips! I'll implement it that way.

Fuco1 commented on org-timeline.el:
```diff
@@ -146,27 +142,56 @@ Return new copy of STRING."
-                (put-text-property start-pos end-pos 'font-lock-face face)
-                (put-text-property start-pos end-pos 'org-timeline-occupied t))
-              (setq current-line 1)))
+              (when (>= beg start-offset)
```
I can't decide if this should be here or in the `org-timeline--list-tasks` function, but I think I would prefer it there so that it returns a definite list for later processing.

Fuco1 reviewed (changes requested):
Thanks for picking up this TODO item, there is a lot of interest in this feature.

I have some comments about the code, especially how to make it more idiomatic and lisp-like.  Feel free to ask additional questions.

Fuco1 commented:
I'm now using https://www.conventionalcommits.org/en/v1.0.0-beta.4/ for the commit messages on all my projects, it would be nice if you could update them.  If you don't know how, don't worry and I'll fix it up before merging (we will need to squash all the commits into one anyway).

aizensoosuke commented:
> I'm now using https://www.conventionalcommits.org/en/v1.0.0-beta.4/ for the commit messages on all my projects, it would be nice if you could update them. If you don't know how, don't worry and I'll fix it up before merging (we will need to squash all the commits into one anyway).

I'm having trouble updating them. Magit only allows editing the last one, and I can't find how to do it on github either. I'll follow this convention from now on, though.

aizensoosuke commented:
> I'm now using https://www.conventionalcommits.org/en/v1.0.0-beta.4/ for the commit messages on all my projects, it would be nice if you could update them. If you don't know how, don't worry and I'll fix it up before merging (we will need to squash all the commits into one anyway).

So, I tried, and failed, to modify them. I'll get more familiar with git and github before continuing. The commits are a mess now, so I'll just let you squash them when you merge. I implemented your other suggestions.

Fuco1 commented:
> I'm having trouble updating them. Magit only allows editing the last one, and I can't find how to do it on github either. I'll follow this convention from now on, though.

You need to do something called `rebase`.  In magit this is done with `r i` while you point on the commit in the history until which you want to change stuff (for example in the log view `l l`).  Then you can put the point on the lines with commits which you want to edit and hit `w` to "reword".

The `r i` opens a separate buffer with a huge list of commands so read through that, there's a lot of things you can do.

Be aware though that this changes the history all the way up to that point and you will need to force push to git to overwrite the old version.

Fuco1 commented on org-timeline.el:
```diff
@@ -148,7 +145,27 @@ Return new copy of STRING."
+                    "|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |"
+                    current-offset))
+         (hourline (org-timeline--add-elapsed-face
+                    (let ((start-hour (abs org-timeline-start-hour)))
```
Just curious why is the `abs` necessary?  What would be the expected use-case of setting this to `-5`?

aizensoosuke replied:
It is, in fact, not necessary at all. I was overzealous about checking user input validy.

Fuco1 replied:
If you really want to validate the input, this is a common pattern in elisp.

1. make a variable `(defcustom package-foo-setting ...)`
2. make a function `(defun package-foo-setting () package-foo-setting)` that just returns the value of the setting
3. use `(package-foo-setting)` to access the setting.
4. if you want to add some validation rules, you can add them to the function accessor now, for example throwing `(user-error "value must be positive number")` if the validation fails.

aizensoosuke replied:
Okay, thank you for that!

aizensoosuke commented:
> You need to do something called `rebase`. In magit this is done with `r i` while you point on the commit in the history until which you want to change stuff (for example in the log view `l l`). Then you can put the point on the lines with commits which you want to edit and hit `w` to "reword".
> 
> The `r i` opens a separate buffer with a huge list of commands so read through that, there's a lot of things you can do.
> 
> Be aware though that this changes the history all the way up to that point and you will need to force push to git to overwrite the old version.

Okay thank you, I'll create a test repo to play around with git and then make the changes here.

aizensoosuke commented:
This should be done now. I'm glad I learned more about git, I'll try to implement more things now. Thank you for your help along the way.

Commit:
feat: tooltip on hover

Commit:
feat: make blocks clickable

Commit:
feat: add block info below timeline on hover

It uses the property 'help-echo which can be used to call a function
when text with the property is called.

Commit:
fix: misusing defvar

Commit:
fix: no more flickering when hovering blocks

Commit:
feat: better info for block hover

instead of just part of the text, hovering blocks now yanks the
whole agenda line (with properties, fonts, everything) in the info
area.

Fuco1 commented:
If you start working on more unrelated changes, it's best to open one pull request for each.  The reasons are pragmatic ones, namely more isolated changes are easier to review.  And also getting the changes merged as fast as possible is generally a good thing, such that only the most minimal amount of work is separated in various branches over time.

Remember that the code in a pull request is not part of the project until it is *merged* (i.e. the owners of the repository approve it and merge it into the mainline code).  Sometimes this means that if you build on top of some changes from a pending pull request, you would need to start your new git branch from that branch instead of main/master branch.

You don't have to split this into multiple PRs now, I'll try to merge it today and then we can open new ones for each feature.

Fuco1 commented on org-timeline.el:
```diff
@@ -62,6 +62,20 @@
+  :type 'integer
+  :group 'org-timeline)
+
+(defconst org-timeline-first-line 0
+  "Computer first line of the timeline in the buffer.")
+
+(defconst org-timeline-height 0
+  "Computed height (number of lines) of the timeline.")
```
If these are computed, I think they shouldn't be marked as constants, since they can change between invocations of agenda.  Only really unchangable things should be marked as constants, the reason is that also the byte compiler can possibly in-line them and produce incorrect code (that is, instead of the variable it will just place the value there which won't change if in the future the variable is re-assigned).

aizensoosuke replied:
Well, I read that "defconst are the least constant variables in elisp". `defvar` doesn't seem to work, and from what I read `defconst` was the way to go for variables that weren't customizable.

aizensoosuke replied:
Here is the [link](http://endlessparentheses.com/what-s-a-defconst-and-why-you-should-use-it.html) to where I read that. I did read too fast.

Although it seems like it wasn't the reason I ran into problems. I'll change those to `defvar` again.

Fuco1 commented on org-timeline.el:
```diff
@@ -147,25 +188,55 @@ Return new copy of STRING."
               (when (>= beg start-offset)
-                (push (list beg end face) tasks)))))))
-    (setq tasks (nreverse tasks))
+                (push (list beg end face txt line) tasks)))))))
```
In some future pull request we should probably change the list to some actual data structure, since it is now growing pretty beg and it might become difficult to remember what data is at which position.  A `plist` or a `defstruct` would be a good choice.

aizensoosuke replied:
I'll do that in the next pull request (there isn't any feature left on the todo list, though).

Fuco1 commented on org-timeline.el:
```diff
@@ -147,25 +188,55 @@ Return new copy of STRING."
-                (put-text-property start-pos end-pos 'font-lock-face face)
-                (put-text-property start-pos end-pos 'org-timeline-occupied t))
+                    (end-pos (get-end-pos current-line end))
+                    (move-to-task-map (make-sparse-keymap))
```
This could probably be defined only once outside of the loop, along with the bindings.

Fuco1 commented on org-timeline.el:
```diff
@@ -147,25 +188,55 @@ Return new copy of STRING."
+                                 'org-timeline-task-line line
+                                  'cursor-sensor-functions '(org-timeline--display-info))))
+                (define-key move-to-task-map [mouse-1] 'org-timeline--move-to-task)
+                (put-text-property start-pos end-pos 'keymap move-to-task-map)
+                (add-text-properties start-pos end-pos props))
```
The keymap can be added to the `props` list, this will let us remove one function call.

aizensoosuke commented:
> If you start working on more unrelated changes, it's best to open one pull request for each. The reasons are pragmatic ones, namely more isolated changes are easier to review. And also getting the changes merged as fast as possible is generally a good thing, such that only the most minimal amount of work is separated in various branches over time.
> 
> Remember that the code in a pull request is not part of the project until it is _merged_ (i.e. the owners of the repository approve it and merge it into the mainline code). Sometimes this means that if you build on top of some changes from a pending pull request, you would need to start your new git branch from that branch instead of main/master branch.
> 
> You don't have to split this into multiple PRs now, I'll try to merge it today and then we can open new ones for each feature.

I wasn't really sure about how to do that, since pushing to branch `aizen` altered this PR. I'll try and do that from now though.

Commit:
perf: move the map definition outside of the loop

Commit:
fix: removing this definition as it is defined above

Commit:
fix: misusing defconst

going back to defvar.

Commit:
fix: marking todo item done

Fuco1 commented:
I would like to finish this PR so we can merge it and you can continue your work on the latest version. 

Can you write a short summary (just a list) of the features that are covered in this pull request?  After that I'll do a final review of the code and we'll merge it in the main line.

Thanks! Super good job so far, especially since you're so new to all this.  I'm very impressed!

aizensoosuke commented:
> Can you write a short summary (just a list) of the features that are covered in this pull request?  After that I'll do a final review of the code and we'll merge it in the main line.

Of course, I'll do that later today though as I don't have access to my computer right now.

> Thanks! Super good job so far, especially since you're so new to all this. I'm very impressed!

Thank you! I really like doing this.

aizensoosuke commented:
Here is a summary of this pull request :
- The hour at which the timeline begins is now customizable with `org-timeline-start-hour`.
- On hover of any block:
    1. The minibuffer displays the corresponding agenda line in the minibuffer. 
    2. A line under the timeline is created, where the same agenda line is displayed. This line never disappears, but it is updated when the user hovers another block. When the user clicks a block (see iii.), this "info line" is removed. This behavious changes considerably in #16 and #17.
    
    Those use the text-property `'help-echo`, which executes a command on hover of the text.
- On `mouse-1` click for any block, the mark is moved to the corresponding agenda line in the agenda buffer.

I want to point out that a lot of things I did here are modified in #16 and #17. Much of the process of displaying info lines changes, a structure is defined for tasks... I think only `org-timeline--move-to-task` doesn't change much.

Fuco1 commented:
Thanks!  I merged this locally after I did some more rebasing (I merged some of the commits together).  Unfortunately github interface doesn't allow to mark this as merged, so I will close the pull request.  However all your work is now included on the main branch, so don't be spooked! :tada:

Fuco1 commented:
For the other work in #16 and #17, you can do a rebase on the `origin/master` to re-use the work which has already been merged.  With a clean working tree (all changes commited or stashed):

```
git fetch origin
git checkout <your-branch>
git rebase origin/master
```

should be enough.

In magit, you can first do `f a` and then `r -A e` and select `origin/master`, or if it is set as upstream, you can substitute `e` for `u`.  The `-A` switch does an automatic stash and unstash in case you have some existing work not yet commited.

Fuco1 commented:
Oh, it seems the function `org-timeline--display-info` is not in this branch, so we broke it :)  Not a big deal as the rebuild on melpa only happens after some time.  Tomorrow I'll go over your other branches and we'll get it fixed in no time.

aizensoosuke commented:
> Oh, it seems the function `org-timeline--display-info` is not in this branch, so we broke it :) Not a big deal as the rebuild on melpa only happens after some time. Tomorrow I'll go over your other branches and we'll get it fixed in no time.

No that's fine, I just tested this branch, there is just an error message (which is not ideal, but it still works). The line that calls `org-timeline--display-info` is obsolete anyway (I removed in #17 I think).

aizensoosuke commented:
> Thanks! I merged this locally after I did some more rebasing (I merged some of the commits together). Unfortunately github interface doesn't allow to mark this as merged, so I will close the pull request. However all your work is now included on the main branch, so don't be spooked! 🎉

Thank you! This is very satisfying, as it is my first true contribution to a project. It has been lot of fun for me.
